File: Node/第七章/06-ai-presentation-agent/server/src/presentation/presentation.types.ts
import * as z from 'zod'

export const EditableRequirementsSchema = z.object({
	topic: z.string().trim().min(2).max(120),
	audience: z.string().trim().min(2).max(120),
	pageCount: z.number().int().min(3).max(12),
	additionalRequirements: z.string().trim().max(1000).default('')
})

export const ThemePresetSchema = z.enum([
	'editorial',
	'technology',
	'business',
	'warm'
])

export const ThemeColorSchema = z
	.string()
	.trim()
	.regex(/^[0-9a-fA-F]{6}$/, '颜色必须是 6 位十六进制色值')
	.transform((value) => value.toUpperCase())

export const ThemeFontSchema = z.enum([
	'Microsoft YaHei',
	'PingFang SC',
	'DengXian',
	'SimHei'
])

export const GeneratedPresentationThemeSchema = z.object({
	name: z.string().trim().min(2).max(20),
	headFontFace: ThemeFontSchema,
	bodyFontFace: ThemeFontSchema,
	coverBackgroundColor: ThemeColorSchema,
	backgroundColor: ThemeColorSchema,
	accentColor: ThemeColorSchema,
	coverTextColor: ThemeColorSchema,
	textColor: ThemeColorSchema,
	coverMutedColor: ThemeColorSchema,
	mutedColor: ThemeColorSchema,
	layoutStyle: z.enum(['side_bar', 'top_line', 'corner_block']),
	titleAlign: z.enum(['left', 'center']),
	density: z.enum(['comfortable', 'compact'])
})

export const GeneratedPageStyleSchema = z.object({
	headFontFace: ThemeFontSchema,
	bodyFontFace: ThemeFontSchema,
	backgroundColor: ThemeColorSchema,
	accentColor: ThemeColorSchema,
	textColor: ThemeColorSchema,
	mutedColor: ThemeColorSchema,
	layoutStyle: z.enum(['side_bar', 'top_line', 'corner_block']),
	titleAlign: z.enum(['left', 'center']),
	density: z.enum(['comfortable', 'compact'])
})

export const CreatePresentationSchema = z.object({
	modelMode: z.enum(['replay', 'ai']).default('replay'),
	...EditableRequirementsSchema.shape,
	sourceText: z.string().trim().min(20).max(60_000),
	sourceName: z.string().trim().min(1).max(160).default('粘贴内容')
})

export const ReviewOutlineSchema = z.object({
	decision: z.enum(['approve', 'revise', 'reject']),
	outlineVersion: z.number().int().positive(),
	feedback: z.string().trim().max(1000).default('')
})

export const RevisePageSchema = z.object({
	changeRequest: z.string().trim().min(2).max(1000)
})

export const ApplyChangeSchema = z.object({
	instruction: z.string().trim().min(2).max(1000)
})

export type CreatePresentationInput = z.infer<typeof CreatePresentationSchema>
export type EditablePresentationRequirements = z.infer<
	typeof EditableRequirementsSchema
>
export type ReviewOutlineInput = z.infer<typeof ReviewOutlineSchema>
export type RevisePageInput = z.infer<typeof RevisePageSchema>
export type ApplyChangeInput = z.infer<typeof ApplyChangeSchema>
export type ThemePreset = z.infer<typeof ThemePresetSchema>
export type GeneratedPresentationTheme = z.infer<
	typeof GeneratedPresentationThemeSchema
>
export type GeneratedPageStyle = z.infer<typeof GeneratedPageStyleSchema>

export type PresentationStatus =
	| 'creating_outline'
	| 'waiting_review'
	| 'generating_pages'
	| 'partially_completed'
	| 'completed'
	| 'rejected'
	| 'exported'

export type OutlineStatus =
	| 'pending_review'
	| 'approved'
	| 'superseded'
	| 'rejected'

export type PageStatus =
	| 'pending'
	| 'generating'
	| 'completed'
	| 'failed'
	| 'stale'

export type PresentationRequirements = Omit<CreatePresentationInput, 'modelMode'>

export interface PresentationTheme {
	preset: ThemePreset | 'custom'
	name: string
	revision: number
	instruction: string | null
	headFontFace: string
	bodyFontFace: string
	coverBackgroundColor: string
	backgroundColor: string
	accentColor: string
	coverTextColor: string
	textColor: string
	coverMutedColor: string
	mutedColor: string
	layoutStyle: 'side_bar' | 'top_line' | 'corner_block'
	titleAlign: 'left' | 'center'
	density: 'comfortable' | 'compact'
}

export interface PresentationChangePlan {
	scope:
		| 'global_content'
		| 'visual_theme'
		| 'single_page'
		| 'single_page_style'
	reason: string
	nextRequirements: EditablePresentationRequirements | null
	themePreset: ThemePreset | null
	generatedTheme: GeneratedPresentationTheme | null
	generatedPageStyle: GeneratedPageStyle | null
	targetPageNumber: number | null
	pageInstruction: string | null
}

export interface PageStyleOverride extends GeneratedPageStyle {
	revision: number
	instruction: string
}

export interface OutlineSlide {
	pageId: string
	order: number
	title: string
	purpose: string
	keyPoints: string[]
}

export interface OutlineVersion {
	version: number
	status: OutlineStatus
	title: string
	slides: OutlineSlide[]
	feedback: string | null
	createdAt: string
	approvedAt: string | null
}

export interface PageArtifact {
	artifactId: string
	outlineVersion: number
	pageRevision: number
	title: string
	subtitle: string
	bullets: string[]
	speakerNote: string
	createdAt: string
}

export interface PageTask {
	pageId: string
	order: number
	title: string
	purpose: string
	outlineVersion: number
	pageRevision: number
	status: PageStatus
	attempts: number
	lastError: string | null
	changeRequest: string | null
	styleOverride: PageStyleOverride | null
	currentArtifactId: string | null
	artifacts: PageArtifact[]
}

export interface TimelineEvent {
	id: string
	type:
		| 'task_created'
		| 'outline_generated'
		| 'outline_revised'
		| 'outline_approved'
		| 'task_rejected'
		| 'page_started'
		| 'page_completed'
		| 'page_failed'
		| 'page_revision_requested'
		| 'page_style_updated'
		| 'requirements_updated'
		| 'theme_updated'
		| 'export_completed'
	message: string
	createdAt: string
}

export interface ExportRecord {
	fileName: string
	filePath: string
	outlineVersion: number
	pageArtifactIds: string[]
	createdAt: string
}

export interface Presentation {
	id: string
	threadId: string
	modelMode: 'replay' | 'ai'
	status: PresentationStatus
	requirements: PresentationRequirements
	theme: PresentationTheme
	outlines: OutlineVersion[]
	currentOutlineVersion: number | null
	pages: PageTask[]
	timeline: TimelineEvent[]
	exportRecord: ExportRecord | null
	createdAt: string
	updatedAt: string
}

export interface OutlineDraft {
	title: string
	slides: Array<Omit<OutlineSlide, 'pageId' | 'order'>>
}

export interface GeneratedPage {
	title: string
	subtitle: string
	bullets: string[]
	speakerNote: string
}

export interface ModelProvider {
	readonly mode: 'replay' | 'ai'
	planChange(input: {
		instruction: string
		presentation: Presentation
	}): Promise<PresentationChangePlan>
	generateOutline(input: {
		requirements: PresentationRequirements
		version: number
		previousOutline: OutlineVersion | null
		feedback: string | null
	}): Promise<OutlineDraft>
	generatePage(input: {
		requirements: PresentationRequirements
		outline: OutlineVersion
		page: PageTask
	}): Promise<GeneratedPage>
}

File: Python/第七章/06-ai-presentation-agent/presentation_types.py
"""AI 演示文稿 Agent 的输入、领域数据和 API 输出模型。"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


ModelMode = Literal["replay", "ai"]
ThemePreset = Literal["editorial", "technology", "business", "warm"]
PresentationStatus = Literal[
    "creating_outline",
    "waiting_review",
    "generating_pages",
    "partially_completed",
    "completed",
    "rejected",
    "exported",
]
OutlineStatus = Literal["pending_review", "approved", "superseded", "rejected"]
PageStatus = Literal["pending", "generating", "completed", "failed", "stale"]
ChangeScope = Literal[
    "global_content",
    "visual_theme",
    "single_page",
    "single_page_style",
]


class CourseModel(BaseModel):
    """统一使用 camelCase，保证 Python API 与课程前端字段完全一致。"""

    model_config = ConfigDict(populate_by_name=True)


class EditableRequirements(CourseModel):
    topic: str = Field(min_length=2, max_length=120)
    audience: str = Field(min_length=2, max_length=120)
    pageCount: int = Field(ge=3, le=12)
    additionalRequirements: str = Field(default="", max_length=1000)

    @field_validator("topic", "audience", "additionalRequirements")
    @classmethod
    def trim_text(cls, value: str) -> str:
        return value.strip()


class CreatePresentationInput(EditableRequirements):
    modelMode: ModelMode = "replay"
    sourceText: str = Field(min_length=20, max_length=60_000)
    sourceName: str = Field(default="粘贴内容", min_length=1, max_length=160)

    @field_validator("sourceText", "sourceName")
    @classmethod
    def trim_source(cls, value: str) -> str:
        return value.strip()


class ReviewOutlineInput(CourseModel):
    decision: Literal["approve", "revise", "reject"]
    outlineVersion: int = Field(gt=0)
    feedback: str = Field(default="", max_length=1000)

    @field_validator("feedback")
    @classmethod
    def trim_feedback(cls, value: str) -> str:
        return value.strip()


class RevisePageInput(CourseModel):
    changeRequest: str = Field(min_length=2, max_length=1000)

    @field_validator("changeRequest")
    @classmethod
    def trim_request(cls, value: str) -> str:
        return value.strip()


class ApplyChangeInput(CourseModel):
    instruction: str = Field(min_length=2, max_length=1000)

    @field_validator("instruction")
    @classmethod
    def trim_instruction(cls, value: str) -> str:
        return value.strip()


class GeneratedPresentationTheme(CourseModel):
    name: str = Field(min_length=2, max_length=20)
    headFontFace: Literal["Microsoft YaHei", "PingFang SC", "DengXian", "SimHei"]
    bodyFontFace: Literal["Microsoft YaHei", "PingFang SC", "DengXian", "SimHei"]
    coverBackgroundColor: str
    backgroundColor: str
    accentColor: str
    coverTextColor: str
    textColor: str
    coverMutedColor: str
    mutedColor: str
    layoutStyle: Literal["side_bar", "top_line", "corner_block"]
    titleAlign: Literal["left", "center"]
    density: Literal["comfortable", "compact"]

    @field_validator(
        "coverBackgroundColor",
        "backgroundColor",
        "accentColor",
        "coverTextColor",
        "textColor",
        "coverMutedColor",
        "mutedColor",
    )
    @classmethod
    def validate_color(cls, value: str) -> str:
        value = value.strip().upper()
        if len(value) != 6 or any(char not in "0123456789ABCDEF" for char in value):
            raise ValueError("颜色必须是 6 位十六进制色值")
        return value


class GeneratedPageStyle(CourseModel):
    headFontFace: Literal["Microsoft YaHei", "PingFang SC", "DengXian", "SimHei"]
    bodyFontFace: Literal["Microsoft YaHei", "PingFang SC", "DengXian", "SimHei"]
    backgroundColor: str
    accentColor: str
    textColor: str
    mutedColor: str
    layoutStyle: Literal["side_bar", "top_line", "corner_block"]
    titleAlign: Literal["left", "center"]
    density: Literal["comfortable", "compact"]

    @field_validator("backgroundColor", "accentColor", "textColor", "mutedColor")
    @classmethod
    def validate_color(cls, value: str) -> str:
        value = value.strip().upper()
        if len(value) != 6 or any(char not in "0123456789ABCDEF" for char in value):
            raise ValueError("颜色必须是 6 位十六进制色值")
        return value


class PresentationTheme(GeneratedPresentationTheme):
    preset: ThemePreset | Literal["custom"]
    revision: int
    instruction: str | None


class PageStyleOverride(GeneratedPageStyle):
    revision: int
    instruction: str


class OutlineSlide(CourseModel):
    pageId: str
    order: int
    title: str
    purpose: str
    keyPoints: list[str] = Field(min_length=2, max_length=5)


class OutlineDraft(CourseModel):
    title: str
    slides: list["OutlineDraftSlide"]


class OutlineDraftSlide(CourseModel):
    title: str
    purpose: str
    keyPoints: list[str] = Field(min_length=2, max_length=5)


class OutlineVersion(CourseModel):
    version: int
    status: OutlineStatus
    title: str
    slides: list[OutlineSlide]
    feedback: str | None
    createdAt: str
    approvedAt: str | None


class GeneratedPage(CourseModel):
    title: str
    subtitle: str
    bullets: list[str] = Field(min_length=2, max_length=5)
    speakerNote: str


class PageArtifact(GeneratedPage):
    artifactId: str
    outlineVersion: int
    pageRevision: int
    createdAt: str


class PageTask(CourseModel):
    pageId: str
    order: int
    title: str
    purpose: str
    outlineVersion: int
    pageRevision: int
    status: PageStatus
    attempts: int
    lastError: str | None
    changeRequest: str | None
    styleOverride: PageStyleOverride | None
    currentArtifactId: str | None
    artifacts: list[PageArtifact]


class TimelineEvent(CourseModel):
    id: str
    type: str
    message: str
    createdAt: str


class ExportRecord(CourseModel):
    fileName: str
    filePath: str
    outlineVersion: int
    pageArtifactIds: list[str]
    createdAt: str


class Presentation(CourseModel):
    id: str
    threadId: str
    modelMode: ModelMode = "replay"
    status: PresentationStatus
    requirements: CreatePresentationInput
    theme: PresentationTheme
    outlines: list[OutlineVersion]
    currentOutlineVersion: int | None
    pages: list[PageTask]
    timeline: list[TimelineEvent]
    exportRecord: ExportRecord | None
    createdAt: str
    updatedAt: str


class PresentationChangePlan(CourseModel):
    scope: ChangeScope
    reason: str
    nextRequirements: EditableRequirements | None
    themePreset: ThemePreset | None
    generatedTheme: GeneratedPresentationTheme | None
    generatedPageStyle: GeneratedPageStyle | None
    targetPageNumber: int | None
    pageInstruction: str | None
